homeassistant/components/solarlog/sensor.py

```diff
@@ -1,133 +1,42 @@
-            self.data["yieldMONTH"] = self.api.yield_month / 1000
-            self.data["yieldYEAR"] = self.api.yield_year / 1000
-            self.data["yieldTOTAL"] = self.api.yield_total / 1000
-            self.data["consumptionAC"] = self.api.consumption_ac
-            self.data["consumptionDAY"] = self.api.consumption_day / 1000
-            self.data["consumptionYESTERDAY"] = self.api.consumption_yesterday / 1000
-            self.data["consumptionMONTH"] = self.api.consumption_month / 1000
-            self.data["consumptionYEAR"] = self.api.consumption_year / 1000
-            self.data["consumptionTOTAL"] = self.api.consumption_total / 1000
-            self.data["totalPOWER"] = self.api.total_power
-            self.data["alternatorLOSS"] = self.api.alternator_loss
-            self.data["CAPACITY"] = round(self.api.capacity * 100, 0)
-            self.data["EFFICIENCY"] = round(self.api.efficiency * 100, 0)
-            self.data["powerAVAILABLE"] = self.api.power_available
-            self.data["USAGE"] = round(self.api.usage * 100, 0)
-            _LOGGER.debug("Updated Solarlog overview data: %s", self.data)
-        except AttributeError:
-            _LOGGER.error("Missing details data in Solarlog response")
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        SolarlogSensor(coordinator, description) for description in SENSOR_TYPES
+    )
 
 
-class SolarlogSensor(SensorEntity):
+class SolarlogSensor(update_coordinator.CoordinatorEntity, SensorEntity):
     """Representation of a Sensor."""
 
+    entity_description: SolarLogSensorEntityDescription
+
     def __init__(
         self,
-        entry_id: str,
-        device_name: str,
-        data: SolarlogData,
+        coordinator: SolarlogData,
         description: SolarLogSensorEntityDescription,
     ) -> None:
         """Initialize the sensor."""
+        super().__init__(coordinator)
         self.entity_description = description
-        self.data = data
-        self._attr_name = f"{device_name} {description.name}"
-        self._attr_unique_id = f"{entry_id}_{description.key}"
+        self._attr_name = f"{coordinator.name} {description.name}"
+        self._attr_unique_id = f"{coordinator.unique_id}_{description.key}"
         self._attr_device_info = {
-            "identifiers": {(DOMAIN, entry_id)},
-            "name": device_name,
+            "identifiers": {(DOMAIN, coordinator.unique_id)},
+            "name": coordinator.name,
             "manufacturer": "Solar-Log",
         }
 
-    def update(self):
-        """Get the latest data from the sensor and update the state."""
-        self.data.update()
-        self._attr_native_value = self.data.data[self.entity_description.json_key]
+    @property
+    def native_value(self) -> StateType:
+        """Return the native sensor value."""
+        return self.coordinator.data[self.entity_description.json_key]
```

Comment:
Side note for the future: If we rename the keys in the data returned by the coordinator to the same names as the entity description keys, we can drop the `json_key` from the entity description.

Reply:
Thanks, But that will cause a change in the unique_id. I don’t think that is desired.

`self._attr_unique_id = f"{coordinator.unique_id}_{description.key}"`

Reply:
No, we won't change the description `key` attribute just remove the `json_key` attribute.

Reply:
But I need the JSON key to get the correct value from the sunwatcher pypi package. This package is having partly capitalized works as keys in its output

Reply:
We don't use the `json_key` to get the data from the pypi library. We use it to get the data from the coordinator, which is data we define in this integration. We can define that data dict however we want.

Reply:
Ah, sorry, you are totally right. I’ll add that in a future PR